extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -298,51 +299,40 @@ export class AmpA4A extends AMP.BaseElement {
             return null;
           }
 
-          // Alias for validCreativePromise's resolve function. This will be called
-          // if any of the fetched keys successfully validate the creative.
-          let resolveValidation;
-          // Promise that will resolve upon any successful validation.
-          const validCreativePromise = new Promise(resolve => {
-            resolveValidation = resolve;
-          });
-
-          // The following block defines an Array of Promises of ArrayBuffers,
-          // none of which will ever resolve. In the first instance of a
-          // successful creative verification, resolveValidation will be
-          // called with the creative as the argument.
-          const signatureVerificationPromises = this.keyInfoSetPromises_.map(
-              keyInfoSetPromise => {
-                return keyInfoSetPromise.then(keyInfoSet => {
-                  return Promise.all(keyInfoSet.map(keyInfoPromise => {
-                    return keyInfoPromise.then(keyInfo => {
-                      if (keyInfo) {
-                        return verifySignature(
-                            new Uint8Array(creativeParts.creative),
-                            creativeParts.signature,
-                            keyInfo)
-                        .then(isValid => {
-                          if (isValid) {
-                            resolveValidation(creativeParts.creative);
-                          }
-                        },
-                        err => {
-                          user().error('Amp Ad', err, this.element);
-                        });
+          // For each signing service, we have exactly one Promise,
+          // keyInfoSetPromise, that holds an Array of Promises of signing keys.
+          // So long as any one of these signing services can verify the
+          // signature, then the creative is valid AMP.
+          return any(this.keyInfoSetPromises_.map(keyInfoSetPromise => {
+            // Resolve Promise into Array of Promises of signing keys.
+            return keyInfoSetPromise.then(keyInfoSet => {
+              // As long as any one individual key of a particular signing
+              // service, keyInfoPromise, can verify the signature, then the
+              // creative is valid AMP.
+              return any(keyInfoSet.map(keyInfoPromise => {
+                // Resolve Promise into signing key.
+                return keyInfoPromise.then(keyInfo => {
+                  if (keyInfo) {
+                    // If the key exists, try verifying with it.
+                    return verifySignature(
+                        new Uint8Array(creativeParts.creative),
+                        creativeParts.signature,
+                        keyInfo)
+                    .then(isValid => {
+                      if (isValid) {
+                        return creativeParts.creative;
                       }
+                      return Promise.reject();
```

Comment:
I don't really mind using Promise.reject in the way that we do here, so all else being equal, I think we should just stick with the approach we currently have. If, in the future, we decide to build an asynchronous iterator for other purposes, then perhaps we could revisit this and see if we can reimplement this bit of code using the iterator.